### PR TITLE
[FrameworkBundle] Allow clearing private cache pools in cache:pool:clear

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

In https://github.com/symfony/symfony/pull/18667/commits/a2a567d1caf3466330b37b74f222db5f006fdfcd cache pools have been passed to `public: false` by default.
The problem is that in 3.2 the `cache:pool:clear` command has been introduced and uses `Container::get($pool);` for clearing them, but the service can't be found since it is private.

So I propose to add a global clearer being able to clear any pool.

### CODE REVIEW

**Comment by Koc** [comment]
❤️  command as service

**Comment by nicolas-grekas** [comment]
Changing the default behavior is a BC break so not possible. Solution 1. is simplest, yet it would make sense to allow clearing any pool with the command I agree. I'd prefer a 3rd option: use the cache pool clearer service to clear by name (method to be added on it to allow so.)
Not a bug though, but a new feature.

**Comment by chalasr** [comment]
@nicolas-grekas I would like to go for your 3rd option but I believe adding this new method will be difficult since the `Psr6CacheClearer` doesn't reference pools by id (see `Psr6CacheClearer::addPool()`). Any clue? Otherwise, solution 2 looks the only working solution for this feature.

Btw, could you please re-label this as a feature? Thanks

**Comment by chalasr** [comment]
Nevermind, it's possible using the `@cache_clear` service.

**Comment by nicolas-grekas** [comment]
Relabeling done.
Not sure about command as a service.
Since each pool can have a different cache clearer, I think we should not reuse the existing cache.default_clearer for that purpose but create a new service (another Psr6CacheClearer enhanced to fit the need?) dedicated to this use case.

**Comment by chalasr** [comment]
Thanks for the hints, let's see what it gives.

Status: needs work

**Comment by chalasr** [comment]
I added a `CachePoolClearer` which is quite similar to the `Psr6CacheClearer` but referencing pools by id, marked as final and internal for avoiding confusions. Not sure it is the best way to do it, but it make us able to clear any private cache pool from its id while keeping the same behavior. 

Status: needs review

**Comment by nicolas-grekas** [comment]
What about this other strategy:
- add a constructor to Psr6CacheClearer that takes an array of cache pools as argument (indexed by pool name)
- deprecated the `add()` method
- add a `clearPool($key)` method (instead of get/has) that returns true/false

Note also that the registration of the pool in the new service must be done in CachePoolClearerPass (not in CachePoolPass), the reason being that otherwise, unused private pools can't be removed by the compiler pass that removes unsued private services.

**Comment by chalasr** [comment]
> What about this other strategy 

Sounds good :)

> CachePoolClearerPass (not in CachePoolPass), the reason being that otherwise, unused private pools can't be removed by the compiler pass that removes unsued private services.

If I move this logic in the `CachePoolClearerPass`, then my private cache pool (which is indeed unused but still exist) is not injected in the new service thus not found by the `cache:pool:clear` command it. Is is the behavior we expect? I thought that any cache pool should be clearable, even unused ones, but maybe it makes no sense

**Comment by nicolas-grekas** [comment]
> is not injected in the new service thus not found

at least it matches the behavior of calling `cache:pool:clear cache.default_clearer` which does not clear unused private cache pools. And it looks sensible to me.

**Comment by chalasr** [comment]
@nicolas-grekas In fact it's already working for private pools which are not "unused", since any private service being referenced at least twice can be retrieved using `Container::get()`.

But to me, it's common to inject a pool only once so I think it would be legit to be able to use the command for clearing it in this case, isn't it? 

> unused private pools can't be optimized by the compiler pass that removes unsued private services.

Note that even after registering pools in the new service from the `CachePoolPass`, if the pool is not referenced at least once (thus really not used), the container still properly remove its definition as it is still considered unused. 

So isn't it doable to keep it at is?

**Comment by nicolas-grekas** [comment]
> any private service being referenced at least twice can be retrieved using Container::get()

That's deprecated so we can't build on that.

I don't get your second point nor why do you want to keep it this way. What would be the goal?

**Comment by chalasr** [comment]
Oh indeed, I forget about this deprecation.

> What would be the goal?

Using the command for clearing a pool being injected in only one service is not possible actually, even with the approach you're proposing since the definition doesn't exist anymore in the `CachePoolClearerPass`. The pool needs to be referenced at least twice for being able to use the command for clearing it, and I thought it could be even better if any pool that is injected (even only once) would be clearable through this command. 
But if we're comfortable with that, fair enough. I implemented your proposal, though I'm not sure to use it the right way, could you give it an eye?

**Comment by chalasr** [comment]
(fabbot doesn't seem to appreciate class names containing digits)

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+imports:
+    - { resource: ../config/default.yml }
+
+services:
+    dummy1:
+        class: Symfony\Bundle\FrameworkBundle\Tests\Fixtures\DeclaredClass
+        arguments: ['@cache.private_pool']
+    dummy2:
+        parent: dummy1
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml]: @nicolas-grekas To clarify, removing this alias make the  `cache.private_pool` unclearable, neither from `cache:clear` nor from `cache:pool:clear`, and it make the `testClearPrivatePool` test fail. I tended to think it's a bug, but maybe I'm wrong

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml]: Understood, it's a bug, I'm going to send a fix asap. Good catch!

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml]: In fact this is not a bug, it's consistent with the fact that you can't clear a private service right now.
Here is the patch that you can apply to fix this:
```diff
diff --git a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolClearerPass.php b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolClearerPass.php
index 00dfdee..20e3da6 100644
--- a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolClearerPass.php
+++ b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolClearerPass.php
@@ -34,7 +34,10 @@ final class CachePoolClearerPass implements CompilerPassInterface
                 if (isset($attr['clearer'])) {
                     $poolsByClearer[$attr['clearer']][$id] = new Reference($id);
                 }
-                if (array_key_exists('clearer', $attr)) {
+                if (!empty($attr['unlazy'])) {
+                    $container->getDefinition($id)->setLazy(false);
+                }
+                if (array_key_exists('clearer', $attr) || array_key_exists('unlazy', $attr)) {
                     break;
                 }
             }
diff --git a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php
index 593b778..4c12a95 100644
--- a/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php
+++ b/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/CachePoolPass.php
@@ -47,8 +47,10 @@ class CachePoolPass implements CompilerPassInterface
             if ($pool->isAbstract()) {
                 continue;
             }
+            $isLazy = $pool->isLazy();
             while ($adapter instanceof DefinitionDecorator) {
                 $adapter = $container->findDefinition($adapter->getParent());
+                $isLazy = $isLazy || $adapter->isLazy();
                 if ($t = $adapter->getTag('cache.pool')) {
                     $tags[0] += $t[0];
                 }
@@ -80,8 +82,16 @@ class CachePoolPass implements CompilerPassInterface
                 throw new InvalidArgumentException(sprintf('Invalid "cache.pool" tag for service "%s": accepted attributes are "clearer", "provider", "namespace" and "default_lifetime", found "%s".', $id, implode('", "', array_keys($tags[0]))));
             }
 
+            $attr = array();
             if (null !== $clearer) {
-                $pool->addTag('cache.pool', array('clearer' => $clearer));
+                $attr['clearer'] = $clearer;
+            }
+            if (!$isLazy) {
+                $pool->setLazy(true);
+                $attr['unlazy'] = true;
+            }
+            if ($attr) {
+                $pool->addTag('cache.pool', $attr);
             }
         }
     }
diff --git a/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml b/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml
index c96183e..393c554 100644
--- a/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml
+++ b/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml
@@ -2,11 +2,9 @@ imports:
     - { resource: ../config/default.yml }
 
 services:
-    dummy1:
+    dummy:
         class: Symfony\Bundle\FrameworkBundle\Tests\Fixtures\DeclaredClass
         arguments: ['@cache.private_pool']
-    dummy2:
-        parent: dummy1
 
 framework:
     cache:
```

File: src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php [hunk 2]
```diff
@@ -55,16 +55,21 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $clearers = array();
         $container = $this->getContainer();
         $cacheDir = $container->getParameter('kernel.cache_dir');
+        $defaultClearer = $container->get('cache.default_clearer');
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php]: you can't use cache.default_clearer here because some cache pool can have a different clearer, so these would be unfetcheable from this clearer.
I propose to add a new `cache.global_clearer` clearer that is unconditionally bound to all pools in CachePoolClearerPass

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php [hunk 3]
```diff
@@ -20,9 +20,32 @@ class Psr6CacheClearer implements CacheClearerInterface
...
     {
-        $this->pools[] = $pool;
+        @trigger_error(sprintf('The %s() method is deprecated since version 3.3 and will be removed in 4.0. Pass an array of pools indexed by name to the constructor instead.', __METHOD__));
+
+        if (!in_array($pool, $this->pools, true)) {
+            $this->pools[] = $pool;
+        }
+    }
+
+    public function hasPool($name)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: This method is useless to me: let's clearPool throw, that's way enough.

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php [hunk 4]
```diff
@@ -20,9 +20,32 @@ class Psr6CacheClearer implements CacheClearerInterface
...
+    {
+        return isset($this->pools[$name]);
+    }
+
+    public function clearPool($name)
+    {
+        if (!isset($this->pools[$name])) {
+            throw new \InvalidArgumentException(sprintf('Cache pool "%s" is not referenced.', $name));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: `Cache pool not found: %s`

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml (hunk 1 above)
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/CachePoolClear/config.yml]: Ok, thank you for the patch then, that's great 👍

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php (hunk 3 above)
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: It's convenient to use it in the command since we want to try fetching the pool from the container if the clearer can't find it. I'll first add the global clearer and come back for this point

**Comment by chalasr** [comment]
@nicolas-grekas changes done, I think we are almost good here, just a doubt about the command itself.

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php [hunk 5]
```diff
@@ -20,9 +20,27 @@ class Psr6CacheClearer implements CacheClearerInterface
...
+    {
+        $this->pools = $pools;
+    }
+
     public function addPool(CacheItemPoolInterface $pool)
     {
-        $this->pools[] = $pool;
+        @trigger_error(sprintf('The %s() method is deprecated since version 3.3 and will be removed in 4.0. Pass an array of pools indexed by name to the constructor instead.', __METHOD__), E_USER_DEPRECATED);
+
+        if (!in_array($pool, $this->pools, true)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: this block is unrelated to the feature of the PR, should be removed

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php [hunk 6]
```diff
@@ -20,9 +20,27 @@ class Psr6CacheClearer implements CacheClearerInterface
...
+        }
+    }
+
+    public function clearPool($name)
+    {
+        if (!isset($this->pools[$name])) {
+            throw new \InvalidArgumentException(sprintf('Cache pool not found: %s.', $name));
+        }
+
+        $this->pools[$name]->clear();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: fir the shake of completeness: `return $this->pools[$name]->clear();`

File: src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php [hunk 7]
```diff
@@ -75,7 +81,12 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         foreach ($pools as $id => $pool) {
             $io->comment(sprintf('Clearing cache pool: <info>%s</info>', $id));
-            $pool->clear();
+
+            if ($pool instanceof CacheItemPoolInterface) {
+                $pool->clear();
+            } else {
+                $globalClearer->clearPool($id);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php]: Now I understand why you needed hasPool :)
This can throw in the middle of a clean because of a missing pool.
Then I propose to replace clearPool($pool) by `PsrCacheClearer::getPools()` and handle validation with the resulting array.
WDYT?

File: src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php (hunk 5 above)
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/CacheClearer/Psr6CacheClearer.php]: It's about keeping BC. If someone rely on `addPool()`, the pool should still be cleared when calling `clear()`, right?  The use of `in_array` is to avoid registering the same pool twice in this case.  Maybe you've a better one?

File: src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php (hunk 7 above)
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php]: Either we keep `hasPool()/clearPool()` or we get all the pools and let the caller doing the filtering logic as you propose, it's debatable but I would prefer keep the clearer being a clearer rather than an accessor. As you prefer :)

Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php]: let's go for hasPool :)

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Command/CachePoolClearCommand.php]: done!

**Comment by nicolas-grekas** [comment]
:+1:
Status: reviewed

**Comment by ogizanagi** [comment]
Either `cache.default_clearer` or `cache.global_clearer` needs to be a `kernel.cache_clearer`, right?

**Comment by nicolas-grekas** [comment]
@ogizanagi which is the case: https://github.com/symfony/symfony/pull/20810/files#diff-1754c5372532c2b324d500472dd5df66R96

**Comment by ogizanagi** [comment]
Indeed. Forgot about the fact tags are not inherited anyway when using a parent definition. 👍

**Comment by ogizanagi** [comment]
However, I found it confusing to reuse the same class for the global cache clearing, even if it's not wired as a kernel cache clearer. What were the arguments against a dedicated `CachePoolRegistry`?

**Comment by chalasr** [comment]
There's no need for accessing pools but only clearing them and like this we can keep pools really private (not accessible from a public service).

**Comment by ogizanagi** [comment]
Ok. Let's move forward. 

👍

**Comment by fabpot** [comment]
Thank you @chalasr.

**Comment by dunglas** [comment]
Maybe that this PR broke tests on master (https://travis-ci.org/symfony/symfony/jobs/183706820) but I don't get exactly what happens (I cannot reproduce in local).

**Comment by chalasr** [comment]
@dunglas tests are broken on 3.2 as well but this is on master only, so it is not the responsible. I tend to think that your first attempt https://github.com/symfony/symfony/pull/20906 was the good, but that tests will stay broken because we released it (maybe I'm wrong, I can't explain it myself).
